vizro-core/src/vizro/_vizro.py

```diff
@@ -144,9 +144,15 @@ def _pre_build():
         # changes size.
         # Any models that are created during the pre-build process *will not* themselves have pre_build run on them.
         # In future may add a second pre_build loop after the first one.
+
+        # model_manager results is wrapped into a list to avoid RuntimeError: dictionary changed size during iteration
+        for _, filter_obj in list(model_manager._items_with_type(Filter)):
+            # Run pre_build on all filters first, then on all other models. This handles dependency between Filter
```

Comment:
> I take it the dependency is that the page needs to know if it is a _dynamic control right? Maybe let's make that clear in the comment 

👍 for this. I did it.

> maybe the page would need to be built before the filter?

My [suggestion](https://github.com/mckinsey/vizro/pull/879#discussion_r1851804902) is opposite 😄. I guess that the parent object should be pre-built after all its children are pre-built. The [Postorder Traversal](https://www.geeksforgeeks.org/postorder-traversal-of-binary-tree/) algorithm will ensure this order. However we can't really know what algorithm will ensure the ideal flow until we spend significant time investigating all the potential dependency combinations.

Reply:
Yes, you are 100% correct. This is a mess and very fragile (not @petar-qb's fault, it's the best we can do for now). I discussed this in https://github.com/mckinsey/vizro/pull/879#discussion_r1848206932 but it probably got lost in all the discussions:

> FYI @maxschulz-COL this is one of the next things I'd like you and me to work on once the pydantic v2 stuff and model manager stuff is done. I have some long term tickets written on it, but basically we need to be able to control the order of pre-build methods much better than we do now. I suspect that any traversal (depth first or otherwise) order won't be flexible enough for us and instead we'll need to do something like pre_build(priority=100) and sort all prebuild methods to execute in the right order. Not having this flexibility has already caused some problems even before this filter one. 

In general we need to figure out a much better way to order our various validation, pre-build etc. operations (see also the example with `controls` needing to come after `components`). At the moment it's very fragile and there's lots of implicit assumptions.